android/src/main/java/io/mosip/tuvali/wallet/WalletBleCommunicator.kt

```diff
@@ -0,0 +1,324 @@
+package io.mosip.tuvali.wallet
+
+import android.bluetooth.BluetoothDevice
+import android.bluetooth.le.ScanRecord
+import android.content.Context
+import android.os.HandlerThread
+import android.os.ParcelUuid
+import android.os.Process
+import android.util.Log
+import io.mosip.tuvali.ble.central.Central
+import io.mosip.tuvali.ble.central.ICentralListener
+import io.mosip.tuvali.common.advertisementPayload.AdvertisementPayload
+import io.mosip.tuvali.common.retrymechanism.BackOffStrategy
+import io.mosip.tuvali.cryptography.SecretsTranslator
+import io.mosip.tuvali.cryptography.WalletCryptoBox
+import io.mosip.tuvali.cryptography.WalletCryptoBoxBuilder
+import io.mosip.tuvali.exception.BLEException
+import io.mosip.tuvali.common.events.VerificationStatusEvent
+import io.mosip.tuvali.common.events.ConnectedEvent
+import io.mosip.tuvali.common.events.DataSentEvent
+import io.mosip.tuvali.common.events.DisconnectedEvent
+import io.mosip.tuvali.common.events.SecureChannelEstablishedEvent
+import io.mosip.tuvali.common.events.EventEmitter
+import io.mosip.tuvali.transfer.TransferReport
+import io.mosip.tuvali.transfer.Util
+import io.mosip.tuvali.transfer.Util.Companion.getLogTag
+import io.mosip.tuvali.verifier.GattService
+import io.mosip.tuvali.verifier.VerifierBleCommunicator
+import io.mosip.tuvali.verifier.VerifierBleCommunicator.Companion.DISCONNECT_STATUS
+import io.mosip.tuvali.wallet.exception.MTUNegotiationFailedException
+import io.mosip.tuvali.wallet.exception.TransferFailedException
+import io.mosip.tuvali.wallet.exception.WalletException
+import io.mosip.tuvali.wallet.transfer.ITransferListener
+import io.mosip.tuvali.wallet.transfer.TransferHandler
+import io.mosip.tuvali.wallet.transfer.message.*
+import org.bouncycastle.util.encoders.Hex
+import java.security.SecureRandom
+import java.util.*
+
+private const val MTU_REQUEST_RETRY_DELAY_TIME_IN_MILLIS = 500L
+
+class WalletBleCommunicator(context: Context, private val eventEmitter: EventEmitter, private val handleException: (BLEException) -> Unit) : ICentralListener, ITransferListener {
+  private val logTag = getLogTag(javaClass.simpleName)
+
+  private val secureRandom: SecureRandom = SecureRandom()
+  private lateinit var verifierPK: ByteArray
+  private var walletCryptoBox: WalletCryptoBox = WalletCryptoBoxBuilder.build(secureRandom)
+  private var secretsTranslator: SecretsTranslator? = null
+
+  private var advPayload: ByteArray? = null
+  private var transferHandler: TransferHandler
+  private val handlerThread = HandlerThread("TransferHandlerThread", Process.THREAD_PRIORITY_DEFAULT)
+
+  private var central: Central
+
+  //default mtu is 23 bytes and the allowed data bytes is 20 bytes
+  private var maxDataBytes = 20
+  private val mtuValues = arrayOf(512, 185, 100)
+
+  private val retryDiscoverServices = BackOffStrategy(maxRetryLimit = 5)
+
+  private enum class CentralCallbacks {
+    ON_DESTROY_SUCCESS_CALLBACK
```

Comment:
We can have this removed since we have eventEmitter injected into this class

Reply:
as we are sending disconnect event conditionally. this can't be removed until we have a better way to sending disconnect event